ddtrace/settings.py

```diff
@@ -118,24 +167,86 @@ class IntegrationConfig(AttrDict):
         config.flask['service_name'] = 'my-service-name'
         config.flask.service_name = 'my-service-name'
     """
-    def __init__(self, global_config, *args, **kwargs):
+    _initialized = False
+
+    def __init__(self, global_config, integration_name, defaults=None):
         """
-        :param global_config:
-        :type global_config: Config
-        :param args:
-        :param kwargs:
+        :param global_config: a reference back to the global config
+        :param integration_name: the name of the integration for this config
+        :param defaults: default key, value pairs to add
+        :type defaults: dict
         """
-        super(IntegrationConfig, self).__init__(*args, **kwargs)
         self.global_config = global_config
         self.hooks = Hooks()
         self.http = HttpConfig()
+        self._integration_name = integration_name
+        self._items = dict()
+
+        # Add the default items to the config if they are specified
+        defaults = defaults or {}
+        for key in defaults:
+            self._add_default(key, defaults[key])
+
+        # This flag is for the __setattr__ to know when we have finished setting
+        # attributes on the class.
+        self._initialized = True
+
+    def __getattr__(self, item):
+        # Use the .items() of the config items (stored in self.items)
+        if item == 'items':
+            return self._items.items
+
+        if item not in self._items:
+            # This will intentionally raise an AttributeError for the config
+            return object.__getattribute__(self, item)
+
+        item = self._items[item]
+        return item.value
+
+    def __getitem__(self, item):
+        return getattr(self, item)
+
+    def __setattr__(self, key, value):
+        # If the class attributes have not yet been initialized or the key is
+        # an attribute on this class then just use the normal object.__setattr__
+        if not self._initialized or hasattr(super(IntegrationConfig, self), key):
+            return object.__setattr__(self, key, value)
+
+        # Else we have either a new attribute or an attribute to be stored as a
+        # config item.
+
+        if not hasattr(self, key):
+            # If the attribute does not exist, add it as a default config item.
+            # TODO: should this be a no-op with a log message instead?
```

Comment:
Thoughts on this? Feels implicit to `_add_default` implicitly when setting an attribute. The user should only be overriding items that already exist and so if we're at this point, it's most likely due to a typo or incorrect usage. A log message seems like the better approach here.